### ISSUE TITLE
Suggestion for get-more generalization

### ISSUE THREAD

**Issue by alexjgriffith** [martianhi/mastodon.el#84]
## Suggestion
Could we generalize some functions in mastodon-tl.el including `mastodon-tl--more-json` and ` mastodon-tl--goto-next-toot`? Currently they only work with `/timeline` endpoints, however `/notifications` and `/accounts` can be processed to appear functionally equivalent. Keeping the buffers functionally equivalent can help make the interaction with each type of buffer consistent.

## Generalizing `mastodon-tl--more-json`
I propose pulling out the URL generation and call into a more general function in mastodon-http.

### Example
In mastodon-http.el we could add `mastodon-http--get-json-with-args` or an equivelent function,
``` elisp
(defun  mastodon-http--get-json-with-args (endpoint args)
  "Return JSON for ENDPOINT with ARGS, where ARGS is a plist."
  (let* ((formated-args (mapconcat
			 (lambda(x)
			   (concat (format "%s" (car x))
				   "="
				   (format "%s" (cdr x))))
			 args "&"))
	 (url (mastodon-http--api (concat endpoint 
					 "?"
					 formated-args))))
    (message url)
    (mastodon-http--get-json url)))

``` 

in mastodon-tl.el `mastodon-tl--more-json` could be rewriten as:
``` elisp
(defun mastodon-tl--more-json(timeline id)
  (mastodon-http--get-json-with-args (concat "timelines/" timeline)
				     `(("max_id".,id))))
```
and  `mastodon-tl--update-json`  as:
``` elisp
(defun mastodon-tl--more-json(timeline id)
  (mastodon-http--get-json-with-args (concat "timelines/" timeline)
				     `(("since_id".,id))))
```
This will let us add versions of more-json for `\accounts` and `\notifications` endpoints.

## Generalizing `mastodon-tl--goto-next-toot`
The best place for generalization is probably in the `mastodon-tl--goto-next-toot` function. This is necessary to call the appropriate more-json function and to provide it with the appropriate argument for the `text-properties` that are relevant in each buffer. 

Currently requesting a next toot `j` calls `mastodon-tl--goto-next-toot` directly as follows:
``` elisp
(defun mastodon-tl--goto-next-toot ()
  "Jump to next toot header."
  (interactive)
  (mastodon-tl--goto-toot-pos 'next-single-property-change
                              'mastodon-tl--more))
```

Some of the options for generalization are replacing `'mastodon-tl--more`  with a `funcall` to a local buffer variable function, or implement minor modes that are implemented for each type of buffer, mapping `j` to `mastodon-<buffer>--goto-next-toot`.

**NOTE:** This also applies to `mastodon-tl--goto-prev-too` and any other functions in mastodon-tl.el that relies on hard coded endpoints.

**Comment by jdenen** [comment]
I'm working on a reply, but in the meantime, can you provide me a general example of how local buffer variable functions would work? I don't think I'm familiar with the idea.

**Comment by alexjgriffith** [comment]
A sort of trivial setup

Set up a function you want to run in buffer a and buffer b. Naming them accordingly is helpful but not necessary
``` elisp
(defun buffer-b--function () (insert "buffer-b\n"))

(defun buffer-a--function () (insert "buffer-a\n"))
```

The following block makes two temp buffer names a-buffer$temp and b-buffer$temp. It switches to buffer a (initializes the buffer) and then sets the local variable `temp-function` and sets its value to `buffer-a--function`. It evaluates the result and then switches to buffer b, where the variable `temp-function` is set to `buffer-b--function`.

Since emacs lisp is a lisp-1 (the variables and functions share the same namespace) you can use funcall and apply on the buffer-variables.

``` elisp
(let ((buffer-a (make-temp-name "a-buffer"))      
      (buffer-b (make-temp-name "b-buffer"))) ;; current buffer
    ;; buffer-a--function acts as it should  
  (switch-to-buffer buffer-a)
  (make-local-variable 'temp-function)
  (setq temp-function 'buffer-a--function)
  (funcall temp-function)
  ;; aliase the function in the current buffer
  (switch-to-buffer buffer-b)
  (make-local-variable 'temp-function)
  (setq temp-function 'buffer-b--function)
  (funcall temp-function)
  ;; buffer-a--function acts as it should
  (switch-to-buffer buffer-a)  
  (funcall temp-function)
  ;; The function is still aliased
  (switch-to-buffer buffer-b)
  (funcall temp-function))
```

Note this is a bit of a hack. There may be a better way of accomplishing this if you want stability and code readability in the long run.

Also, `buffer-a--function` or `buffer-b--function` could return a lambda, meaning the current code would not be impacted.

example
``` elisp
(defun buffer-a--function () (lambda (arg-1 arg-2) ... ))
```

## How it would look
Place this in mastodon.el or something like mastodon-locals.el as a back up in case the local function is not set.
``` elisp
(defvar mastodon--get-more nil)
```

nil could be replaced with a function with the same arity
``` elisp
(defvar mastodon--get-more
  (lambda() (lambda(var1 var2 ...) 
              (message
               "mastodon--get-more not defined for current-buffer") )) 
```

Call this in initiation of any timeline buffer
``` elisp
(defun mastodon-tl--init()
   ;; alias global function
  (make-local-variable 'mastodon--get-more)
  (setq mastodon--get-more
        (lambda ()(lambda(var1 var2 ...) 
                    (mastodon-tl--get-more var1 var2 ...))))
```

This can be kept in the mastodon-tl.el or moved to a module such as mastodon-locals.el so it is clear that it relies on a local function.
``` elisp
(defun mastodon-tl--goto-next-toot ()
  "Jump to next toot header."
  (interactive)
  (mastodon-tl--goto-toot-pos 'next-single-property-change
                              (funcall mastodon--get-more))) ;; no ' before the function
```

**Comment by jdenen** [comment]
Yeah, I'd prefer to go the readability route. That seems clever but it is hard to understand.

When I started down this road, I envisioned a minor-mode per timeline, with `mastodon-tl.el` being a library of macros to define the minor modes. Then a `mastodon-timelines.el` file with definitions for home/local/public/tags since they all behave basically the same. 

Notifications and profiles would get their own features, but use the `tl` macros as much as possible.

For example (and this is untested semi-pseudo code):

```elisp
;; mastodon-tl.el
(defmacro mastodon-tl--goto-next (api &optional display)
  "Define a navigate-to-next function for API buffer. 

Name buffer functions with DISPLAY if non-nil."
  (let* ((name (or display api))
         (func (intern (format "mastodon-%s--goto-next" name)))
         (docs (format "Move to next item in %s buffer." name)))
    `(defun ,func () ,docs
            (mastodon-tl--goto-toot-pos 'next-single-property-change
                                        ;; assume a macro defines this function elsewhere
                                        ;; but it can be custom as needed
                                        (intern (format "mastodon-%s--more" name))))))
```

```elisp
;; mastodon-timelines.el
(mastodon-tl--goto-next "timelines/home" "home")
(mastodon-tl--goto-next "timelines/public" "federated")
(mastodon-tl--goto-next "timelines/public?local=true" "local")
```

Then the major mode keymap contains behavior like switching between timelines, sending new toots, etc. And the minor mode keymaps should be defined by a macro too.

---

This is all what I had tentatively in my head toward the beginning of major mode development. It's why `mastodon-tl` is not `mastodon-timeline`, for example. But then I wanted to rush to push features and cross this bridge when I came to it.

Then you started contributing and we came to it much sooner than I thought. 😎

**Comment by alexjgriffith** [comment]
That seams far more maintainable! 

I can see how it nicely sets up sepereate movement functions for each buffer type.

Just to clarify, currently the `tl-get`set of functions are bound to keys in the major-mode. You planing on having one minor mode for each type of buffer and binding type     specific`tl-get`functions to each? For example, `notifications-get-next` in a notifications buffer, `local-get-next` in a local buffer, `global-get-next` in the global buffer etc.

**Comment by alexjgriffith** [comment]
Is there an idomatic way for the user to set keybindings for a set of minor buffers?

For example, could I change my binding from `j` in all 5+ minor modes to `C-c j`?

**Comment by jdenen** [comment]
```

**Comment by jdenen** [comment]
I believe something like this will work:
```elisp
;; mastodon.el
(defcustom mastodon-buffer-next-key "j"
  "Keybinding to move to next item in a Mastodon buffer.")

;; mastodon-tl.el
(defmacro mastodon-tl--setup-kbds (timeline)
  "Setup keybindings for TIMELINE."
  (let ((map (intern (format "mastodon-%s-mode-map" timeline))))
    `((define-key map ,(kbd mastodon-buffer-next-key) #',(intern (format "mastodon-%s--goto-next" timeline))))))

;; elsewhere
(mastodon-tl--setup-kbds "home")
(mastodon-tl--setup-kbds "local")
(mastodon-tl--setup-kbds "notifications")
```
It's customization for ALL timeline buffers at once, but I think that's OK.

**Comment by alexjgriffith** [comment]
That is beautiful.

All the user would have to do to change keybindings would be to add the following to their init buffer.

``` elisp
(require 'mastodon)
(customize-set-variable mastodon-buffer-next-key "C-c j")
(mastodon-tl--setup-kbds "home")
;; or another function that does it for a set of similar
;; minor modes
```

Edit: fixed phone typos

**Comment by jdenen** [comment]
Users shouldn't have to use the `setup-kbds` macro. We'd call that in one of the mastodon features. Users just have to tweak the `-next-key` variable in their `.emacs` or use `customize` interactively.

But it's late and I may be misunderstanding what you mean by "their unit buffer."

**Comment by jdenen** [comment]
Ah, init buffer makes much more sense 😆 

Yeah, I don't think they'd need to call the macro, but I'd have to prove that out before I was certain.

**Comment by alexjgriffith** [comment]
Yep, the macro will only ever have to be called once! For a moment I forgot that everything in emacs is a global.

**Comment by hdurer** [comment]
I am not convinced that having macros that magically define new functions is more readable.
What is wrong with helper functions and very thin actual function?  Sure, there is a bit of repetition but user can easily find the source of a function's definition and see the actual implementation.

**Comment by jdenen** [comment]
I'm inclined to agree with you. As I tinker locally, the macro solution is becoming increasingly difficult to follow in my own head. I can only image what it would be like for someone who is not writing it.

**Comment by alexjgriffith** [comment]
Taking a look at `mastodon-tl--updated-json`, `mastodon-tl--update` and `mastodon-tl--timeline-name` I can see a simple solution to the generalization problem.

We are already using the `buffer-name` to identify which timeline we are targeting with `-update` and `-more`. We can either change the `buffer` name to identify what type of endpoint we are targeting (\*mastodon-tl-home\* \*mastodon-notifications\*) and adjust `mastodon-tl--updated-json` to select the specific case, or, make `mastodon-tl--timeline-name` access our own buffer local variable instead of reading the `buffer-name` buffer local variable.

``` elisp
(defun mastodon-tl--updated-json (timeline id)
  "Return JSON for TIMELINE since ID."
  (let ((url (mastodon-http--api (concat "timelines/"
                                         timeline
                                         "?since_id="
                                         (number-to-string id)))))
    (mastodon-http--get-json url)))
```

``` elisp
(defun mastodon-tl--update ()
  "Update timeline with new toots."
  (interactive)
  (let* ((tl (mastodon-tl--timeline-name))
         (id (mastodon-tl--newest-id))
         (json (mastodon-tl--updated-json tl id)))
    (when json
      (with-current-buffer (current-buffer)
        (let ((inhibit-read-only t))
          (goto-char (point-min))
          (mastodon-tl--timeline json))))))

```

``` elsip
(defun mastodon-tl--timeline-name ()
  "Determine timeline from `buffer-name'."
  (replace-regexp-in-string "\*" ""
                            (replace-regexp-in-string "mastodon-" "" (buffer-name))))
```

**Comment by alexjgriffith** [comment]
Closed by #122
